src/libopensc/pkcs15-esteid2025.c

```diff
@@ -0,0 +1,203 @@
+/*
+ * PKCS15 emulation layer for EstEID card issued from December 2025.
+ *
+ * Copyright (C) 2024, Raul Metsma <raul@metsma.ee>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <ctype.h>
+#include <stdlib.h>
+
+#include "common/compat_strlcpy.h"
+
+#include "internal.h"
+#include "opensc.h"
+#include "pkcs15.h"
+
+static int
+sc_pkcs15emu_esteid2025_init(sc_pkcs15_card_t *p15card)
+{
+	sc_card_t *card = p15card->card;
+	u8 *buff;
+	int r, i;
+	size_t field_length = 0, buflen = 9, j;
+	sc_path_t tmppath;
+	static const u8 prkey_id[2] = {0x01, 0x02};
+
+	set_string(&p15card->tokeninfo->label, "ID-kaart");
+	set_string(&p15card->tokeninfo->manufacturer_id, "Thales");
+
+	buff = malloc(buflen + 1);
+	if (!buff)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+
+	/* Read documber number to be used as serial */
+	sc_format_path("DFDD5007", &tmppath);
+	LOG_TEST_RET(card->ctx, sc_select_file(card, &tmppath, NULL), "SELECT docnr");
+	r = sc_read_binary(card, 0, buff, buflen, 0);
+	LOG_TEST_RET(card->ctx, r, "read document number failed");
+
+	for (j = 0; j < buflen; j++)
+		if (!isalnum(buff[j]))
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+	buff[buflen] = '\0';
+
+	free(p15card->tokeninfo->serial_number);
+	p15card->tokeninfo->serial_number = (char *)buff;
+	p15card->tokeninfo->flags = SC_PKCS15_TOKEN_READONLY;
+
+	/* add certificates */
+	for (i = 0; i < 2; i++) {
+		static const char *cert_names[2] = {"Isikutuvastus", "Allkirjastamine"};
+		static const char *cert_paths[2] = {"ADF1:3411", "ADF2:3421"};
+
+		struct sc_pkcs15_cert_info cert_info = {
+				.id = {.len = 1, .value[0] = prkey_id[i]}
+		 };
```

Comment:
nit: Would prefer to move this outside of the cycle, or at least outside of the function